File: client/src/models/solver.js
export class Solver {
  constructor(data) {
    this.apiVersion = data.apiVersion || '';
    this.kind = data.kind || '';
    this.metadata = new Metadata(data.metadata);
    this.spec = new Spec(data.spec);
    this.status = new Status(data.status);
  }
}

class Metadata {
  constructor(data) {
    this.annotations = data.annotations || {};
    this.creationTimestamp = data.creationTimestamp || '';
    this.generation = data.generation || 0;
    this.managedFields = (data.managedFields || []).map(field => new ManagedField(field));
    this.name = data.name || '';
    this.namespace = data.namespace || '';
    this.resourceVersion = data.resourceVersion || '';
    this.uid = data.uid || '';
  }
}

class ManagedField {
  constructor(data) {
    this.apiVersion = data.apiVersion || '';
    this.fieldsType = data.fieldsType || '';
    this.fieldsV1 = data.fieldsV1 || {};
    this.manager = data.manager || '';
    this.operation = data.operation || '';
    this.time = data.time || '';
  }
}

class Spec {
  constructor(data) {
    this.establishPeering = data.establishPeering || false;
    this.findCandidate = data.findCandidate || false;
    this.intentID = data.intentID || '';
    this.reserveAndBuy = data.reserveAndBuy || false;
    this.selector = new Selector(data.selector);
  }
}

class Selector {
  constructor(data) {
    this.filters = new Filters(data.filters);
    this.flavorType = data.flavorType || '';
  }
}

class Filters {
  constructor(data) {
    this.architectureFilter = new Filter(data.architectureFilter);
    this.cpuFilter = new Filter(data.cpuFilter);
    this.memoryFilter = new Filter(data.memoryFilter);
    this.podsFilter = new Filter(data.podsFilter);
  }
}

class Filter {
  constructor(data) {
    this.data = new FilterData(data.data);
    this.name = data.name || '';
  }
}

class FilterData {
  constructor(data) {
    this.value = data.value || '';
    this.min = data.min || '';
    this.max = data.max || '';
  }
}

class Status {
  constructor(data) {
    this.allocation = new ResourceRef(data.allocation);
    this.contract = new ResourceRef(data.contract);
    this.credentials = new Credentials(data.credentials);
    this.discoveryPhase = data.discoveryPhase || '';
    this.findCandidate = data.findCandidate || '';
    this.peering = data.peering || '';
    this.reservationPhase = data.reservationPhase || '';
    this.reserveAndBuy = data.reserveAndBuy || '';
    this.solverPhase = new SolverPhase(data.solverPhase);
  }
}

class ResourceRef {
  constructor(data) {
    this.name = data.name || '';
    this.namespace = data.namespace || '';
  }
}

class Credentials {
  constructor(data) {
    this.clusterID = data.clusterID || '';
    this.clusterName = data.clusterName || '';
    this.endpoint = data.endpoint || '';
    this.token = data.token || '';
  }
}

class SolverPhase {
  constructor(data) {
    this.endTime = data.endTime || '';
    this.lastChangeTime = data.lastChangeTime || '';
    this.message = data.message || '';
    this.phase = data.phase || '';
  }
}

File: server/models/solver.go
package models

import "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"

type Filter struct {
	Mode  string `json:"mode"`
	Min   string `json:"min,omitempty"`
	Max   string `json:"max,omitempty"`
	Value string `json:"value,omitempty"`
}

type SolverData struct {
	Name               string `json:"name"`
	Type               string `json:"type,omitempty"`
	IntentID           string `json:"intentID"`
	ArchitectureFilter Filter `json:"architectureFilter,omitempty"`
	CPUFilter          Filter `json:"cpuFilter,omitempty"`
	MemoryFilter       Filter `json:"memoryFilter,omitempty"`
	PodsFilter         Filter `json:"podsFilter,omitempty"`
	FindCandidate      bool   `json:"findCandidate"`
	ReserveAndBuy      bool   `json:"reserveAndBuy"`
	EstablishPeering   bool   `json:"establishPeering"`
}

func CreateSolverResource(solverData *SolverData) *unstructured.Unstructured {
	solverResource := map[string]interface{}{
		"apiVersion": "nodecore.fluidos.eu/v1alpha1",
		"kind":       "Solver",
		"metadata": map[string]interface{}{
			"name":      solverData.Name,
			"namespace": "fluidos",
		},
		"spec": map[string]interface{}{
			"intentID":         solverData.IntentID,
			"findCandidate":    solverData.FindCandidate,
			"reserveAndBuy":    solverData.ReserveAndBuy,
			"establishPeering": solverData.EstablishPeering,
		},
	}

	selector := map[string]interface{}{}

	// Add field only if they are present. This section should be updated to supports other FlavorType
	if solverData.Type != "" {
		selector["flavorType"] = solverData.Type
	}

	filters := map[string]interface{}{}
	if solverData.ArchitectureFilter.Mode != "" {
		filters["architectureFilter"] = map[string]interface{}{
			"name": solverData.ArchitectureFilter.Mode,
			"data": createFilterData(&solverData.ArchitectureFilter),
		}
	}
	if solverData.CPUFilter.Mode != "" {
		filters["cpuFilter"] = map[string]interface{}{
			"name": solverData.CPUFilter.Mode,
			"data": createFilterData(&solverData.CPUFilter),
		}
	}
	if solverData.MemoryFilter.Mode != "" {
		filters["memoryFilter"] = map[string]interface{}{
			"name": solverData.MemoryFilter.Mode,
			"data": createFilterData(&solverData.MemoryFilter),
		}
	}
	if solverData.PodsFilter.Mode != "" {
		filters["podsFilter"] = map[string]interface{}{
			"name": solverData.PodsFilter.Mode,
			"data": createFilterData(&solverData.PodsFilter),
		}
	}

	if len(filters) > 0 {
		selector["filters"] = filters
	}

	if len(selector) > 0 {
		solverResource["spec"].(map[string]interface{})["selector"] = selector
	}

	return &unstructured.Unstructured{Object: solverResource}
}

// Helper to create filter data. Pay attention to server side validation
func createFilterData(filter *Filter) map[string]interface{} {
	data := make(map[string]interface{})

	if filter.Mode == "Match" {
		data["value"] = filter.Value
	} else if filter.Mode == "Range" {
		if filter.Min != "" {
			data["min"] = filter.Min
		}
		if filter.Max != "" {
			data["max"] = filter.Max
		}
	}

	return data
}
